packages/block-library/src/query-loop/edit.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, useMemo } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+import {
+	BlockContextProvider,
+	InnerBlocks,
+	BlockPreview,
+} from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import { useQueryContext } from '../query';
+
+const TEMPLATE = [ [ 'core/post-title' ], [ 'core/post-content' ] ];
+export default function QueryLoopEdit( { clientId, context: { query } } ) {
+	const [ { page } ] = useQueryContext();
+	const [ activeBlockContext, setActiveBlockContext ] = useState();
+
+	const { posts, blocks } = useSelect(
+		( select ) => ( {
+			posts: select( 'core' ).getEntityRecords( 'postType', 'post', {
+				...query,
+				offset: query.per_page * ( page - 1 ) + query.offset,
+				page,
+			} ),
+			blocks: select( 'core/block-editor' ).getBlocks( clientId ),
+		} ),
+		[ query, page, clientId ]
+	);
+
+	const blockContexts = useMemo(
+		() =>
+			posts?.map( ( post ) => ( {
+				postType: post.type,
+				postId: post.id,
+			} ) ),
+		[ posts ]
+	);
+	return blockContexts
+		? blockContexts.map( ( blockContext ) => (
+				<BlockContextProvider
+					key={ blockContext.postId }
+					value={ blockContext }
+				>
+					{ blockContext ===
+					( activeBlockContext || blockContexts[ 0 ] ) ? (
+						<InnerBlocks template={ TEMPLATE } />
```

Comment:
I'd have expect this to be a "controlled" InnerBlocks and the template being saved to an attribute.
The reason being that if we save the template as InnerBlocks, the markup saved (fallback) is not a valid representation of what we want to show, it's just one post.

Reply:
Wouldn't it save the actual template representation (post-title, post-content, etc), not a specific post?

Reply:
yes, it does. I guess maybe it's fine since most of these are dynamic blocks. But you can end up with groups...

Reply:
Yeah, it just saves the actual template representation.